File: lib/node_grab/sablier/ctime.rb
require "rubygems"
require "ffi"

module Ctime
  extend FFI::Library
  ffi_lib "/home/matteo/sviluppo/rqueue_jruby/lib/node_grab/sablier/libc/libctime.so"
  attach_function :getWallTime, [], :long
  attach_function :getCpuTime, [], :double
  attach_function :calcWallTime, [:long,:long], :long
  attach_function :calCpuTime, [:int,:int], :double
#  attach_function :error_code, [], :int # note empty array for functions taking zero arguments
#  attach_function :create_object, [:string], :pointer
#  attach_function :calculate_something_else, [:double, :pointer], :double
#  attach_function :free_object, [:pointer], :void
end

File: lib/node_grab/sablier/libc/ctime.c

#include <sys/types.h>
#include <time.h>
#include <stdio.h>


long getWallTime()
{
	long t;
	t = time(NULL);
	return t;


}

double getCpuTime(){
	int c;
	c = clock();
	return c;
}


long calcWallTime(long t0,long t1){
    long cp ;
    cp = (long) (t1 - t0);
    return cp;

}

double calCpuTime(int c0, int c1){
	double wt;
        wt = (double) (c1 - c0)/CLOCKS_PER_SEC;
	return wt;

}
